// react-auth-kit/RxTokenObject.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _jsCookie = _interopRequireDefault(require("js-cookie"));
var _rxjs = require("rxjs");
var _errors = require("./errors");
class TokenObject {
  constructor(authStorageName, authStorageType, refreshTokenName, cookieDomain, cookieSecure) {
    this.authStorageName = authStorageName;
    this.authStorageType = authStorageType;
    this.stateStorageName = `${authStorageName}_state`;
    this.refreshTokenName = refreshTokenName;
    this.cookieDomain = cookieDomain;
    this.cookieSecure = cookieSecure;
    this.authStorageTypeName = `${this.authStorageName}_type`;
    this.isUsingRefreshToken = !!this.refreshTokenName;
    this.authValue = this.initialToken_();
    this.authSubject = new _rxjs.BehaviorSubject(this.authValue);
    this.authSubject.subscribe({
      next: this.syncTokens,
      complete: () => {
        console.log('Token Synced');
      },
      error: err => {
        console.error('Error Occured while syncing token');
        console.log(err);
      }
    });
  }
  subscribe = (next, error, complete) => {
    this.authSubject.subscribe({
      next: next,
      error: error,
      complete: complete
    });
  };
  observe = () => {
    return this.authSubject.asObservable();
  };
  set = data => {
    let obj = this.value;
    if (data.auth) {
      let userState = obj.userState;
      if (data.userState !== undefined) {
        userState = data.userState;
      }
      obj = {
        ...obj,
        auth: {
          'token': data.auth.token,
          'type': data.auth.type,
          'expiresAt': this.getExpireDateTime(data.auth.token)
        },
        isSignIn: true,
        userState: userState
      };
    } else if (data.auth === null) {
      obj = {
        ...obj,
        auth: null,
        isSignIn: false,
        userState: null
      };
    }
    if (this.isUsingRefreshToken) {
      if (data.refresh) {
        obj = {
          ...obj,
          refresh: {
            'token': data.refresh,
            'expiresAt': this.getExpireDateTime(data.refresh)
          }
        };
      } else if (data.refresh === null) {
        obj = {
          ...obj,
          refresh: null
        };
      }
    }
    this.authValue = obj;
    this.authSubject.next(obj);
  };
  get value() {
    return this.authSubject.value;
  }
  initialToken_ = () => {
    if (this.authStorageType === 'cookie') {
      return this.initialCookieToken_();
    } else {
      return this.initialLSToken_();
    }
  };
  initialCookieToken_ = () => {
    const authToken = _jsCookie.default.get(this.authStorageName);
    const authTokenType = _jsCookie.default.get(this.authStorageTypeName);
    const stateCookie = _jsCookie.default.get(this.stateStorageName);
    const refreshToken = this.isUsingRefreshToken && this.refreshTokenName != null ? _jsCookie.default.get(this.refreshTokenName) : null;
    return this.checkTokenExist_(authToken, authTokenType, stateCookie, refreshToken);
  };
  initialLSToken_ = () => {
    const authToken = localStorage.getItem(this.authStorageName);
    const authTokenType = localStorage.getItem(this.authStorageTypeName);
    const stateCookie = localStorage.getItem(this.stateStorageName);
    const refreshToken = this.isUsingRefreshToken && this.refreshTokenName != null ? localStorage.getItem(this.refreshTokenName) : null;
    return this.checkTokenExist_(authToken, authTokenType, stateCookie, refreshToken);
  };
  checkTokenExist_ = (authToken, authTokenType, stateCookie, refreshToken) => {
    try {
      let refresh;
      if (this.isUsingRefreshToken && !!refreshToken) {
        const refreshTokenExpiresAt = this.getExpireDateTime(refreshToken);
        if (refreshTokenExpiresAt < new Date()) {
          refresh = null;
        } else {
          refresh = {
            token: refreshToken,
            expiresAt: refreshTokenExpiresAt
          };
        }
      } else {
        refresh = null;
      }
      if (this.isUsingRefreshToken && !refresh) {
        this.removeAllToken();
        return {
          auth: null,
          refresh: null,
          userState: null,
          isUsingRefreshToken: this.isUsingRefreshToken,
          isSignIn: false
        };
      }
      let auth;
      let authState;
      if (!!authToken && !!authTokenType && !!stateCookie) {
        try {
          const expiresAt = this.getExpireDateTime(authToken);
          if (expiresAt < new Date()) {
            auth = null;
            authState = null;
          } else {
            authState = JSON.parse(stateCookie);
            auth = {
              token: authToken,
              type: authTokenType,
              expiresAt: expiresAt
            };
          }
        } catch (e) {
          auth = null;
          authState = null;
        }
      } else {
        auth = null;
        authState = null;
      }
      if (refresh) {
        if (!!auth && !!authState) {
          return {
            auth: auth,
            refresh: refresh,
            userState: authState,
            isUsingRefreshToken: this.isUsingRefreshToken,
            isSignIn: true
          };
        }
        this.removeAuth();
        return {
          auth: null,
          refresh: refresh,
          userState: null,
          isUsingRefreshToken: this.isUsingRefreshToken,
          isSignIn: false
        };
      } else if (!this.isUsingRefreshToken && !!auth && !!authState) {
        return {
          auth: auth,
          refresh: null,
          userState: authState,
          isUsingRefreshToken: this.isUsingRefreshToken,
          isSignIn: true
        };
      }
      {
        this.removeAllToken();
        return {
          auth: null,
          refresh: null,
          userState: null,
          isUsingRefreshToken: this.isUsingRefreshToken,
          isSignIn: false
        };
      }
    } catch (e) {
      this.removeAllToken();
      return {
        auth: null,
        refresh: null,
        userState: null,
        isUsingRefreshToken: this.isUsingRefreshToken,
        isSignIn: false
      };
    }
  };
  parseJwt = token => {
    const base64Url = token.split('.')[1];
    const base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
    const jsonPayload = decodeURIComponent(atob(base64).split('').map(function (c) {
      return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
    }).join(''));
    return JSON.parse(jsonPayload);
  };
  getExpireDateTime = token => {
    const jwtData = this.parseJwt(token);
    if (jwtData.hasOwnProperty('exp')) {
      const d = new Date(0);
      d.setUTCSeconds(jwtData.exp);
      return d;
    } else {
      throw new _errors.AuthError('JWT has no exp param');
    }
  };
  syncTokens = authState => {
    if (authState.auth) {
      this.setAuthToken(authState.auth.token, authState.auth.type, authState.userState);
    } else {
      this.removeAuth();
    }
    if (!!authState.refresh && this.isUsingRefreshToken) {
      this.setRefreshToken(authState.refresh.token);
    } else {
      this.removeRefresh();
    }
  };
  setAuthToken = (authToken, authTokenType, authState) => {
    if (this.authStorageType === 'cookie') {
      const expiresAt = this.getExpireDateTime(authToken);
      _jsCookie.default.set(this.authStorageName, authToken, {
        expires: expiresAt,
        domain: this.cookieDomain,
        secure: this.cookieSecure
      });
      _jsCookie.default.set(this.authStorageTypeName, authTokenType, {
        expires: expiresAt,
        domain: this.cookieDomain,
        secure: this.cookieSecure
      });
      if (authState) {
        _jsCookie.default.set(this.stateStorageName, JSON.stringify(authState), {
          expires: expiresAt,
          domain: this.cookieDomain,
          secure: this.cookieSecure
        });
      }
    } else {
      window.localStorage.setItem(this.authStorageName, authToken);
      window.localStorage.setItem(this.authStorageTypeName, authTokenType);
      if (authState) {
        window.localStorage.setItem(this.stateStorageName, JSON.stringify(authState));
      }
    }
  };
  setRefreshToken = refreshToken => {
    if (this.authStorageType === 'cookie') {
      if (this.isUsingRefreshToken && !!this.refreshTokenName && !!refreshToken) {
        const refreshTokenExpiresAt = this.getExpireDateTime(refreshToken);
        _jsCookie.default.set(this.refreshTokenName, refreshToken, {
          expires: refreshTokenExpiresAt,
          domain: this.cookieDomain,
          secure: this.cookieSecure
        });
      }
    } else {
      if (this.isUsingRefreshToken && !!this.refreshTokenName && !!refreshToken) {
        localStorage.setItem(this.refreshTokenName, refreshToken);
      }
    }
  };
  removeAllToken = () => {
    if (this.authStorageType === 'cookie') {
      this.removeAllCookieToken_();
    } else {
      this.removeAllLSToken_();
    }
  };
  removeAllCookieToken_ = () => {
    _jsCookie.default.remove(this.authStorageName, {
      domain: this.cookieDomain,
      secure: this.cookieSecure
    });
    _jsCookie.default.remove(this.authStorageTypeName, {
      domain: this.cookieDomain,
      secure: this.cookieSecure
    });
    _jsCookie.default.remove(this.stateStorageName, {
      domain: this.cookieDomain,
      secure: this.cookieSecure
    });
    if (this.isUsingRefreshToken && !!this.refreshTokenName) {
      _jsCookie.default.remove(this.refreshTokenName, {
        domain: this.cookieDomain,
        secure: this.cookieSecure
      });
    }
  };
  removeAllLSToken_ = () => {
    localStorage.removeItem(this.authStorageName);
    localStorage.removeItem(this.authStorageTypeName);
    localStorage.removeItem(this.stateStorageName);
    if (this.isUsingRefreshToken && !!this.refreshTokenName) {
      localStorage.removeItem(this.refreshTokenName);
    }
  };
  removeAuth = () => {
    if (this.authStorageType === 'cookie') {
      this.removeAuthCookie();
    } else {
      this.removeAuthToken();
    }
  };
  removeAuthCookie = () => {
    _jsCookie.default.remove(this.authStorageName, {
      domain: this.cookieDomain,
      secure: this.cookieSecure
    });
    _jsCookie.default.remove(this.authStorageTypeName, {
      domain: this.cookieDomain,
      secure: this.cookieSecure
    });
    _jsCookie.default.remove(this.stateStorageName, {
      domain: this.cookieDomain,
      secure: this.cookieSecure
    });
  };
  removeAuthToken = () => {
    localStorage.removeItem(this.authStorageName);
    localStorage.removeItem(this.authStorageTypeName);
    localStorage.removeItem(this.stateStorageName);
  };
  removeRefresh = () => {
    if (this.authStorageType === 'cookie') {
      this.removeRefreshCookie();
    } else {
      this.removeRefreshLocalStorage();
    }
  };
  removeRefreshCookie = () => {
    if (this.isUsingRefreshToken && !!this.refreshTokenName) {
      _jsCookie.default.remove(this.refreshTokenName, {
        domain: this.cookieDomain,
        secure: this.cookieSecure
      });
    }
  };
  removeRefreshLocalStorage = () => {
    if (this.isUsingRefreshToken && !!this.refreshTokenName) {
      localStorage.removeItem(this.refreshTokenName);
    }
  };
}
var _default = exports.default = TokenObject;
