// react-auth-kit/utils/reducers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.doRefresh = doRefresh;
exports.doSignIn = doSignIn;
exports.doSignOut = doSignOut;
function doSignIn(signInParams) {
  const authType = signInParams.auth.type ? signInParams.auth.type : 'Bearer';
  const authToken = signInParams.auth.token;
  return {
    auth: {
      token: authToken,
      type: authType
    },
    refresh: signInParams.refresh,
    userState: signInParams.userState || {}
  };
}
function doRefresh(refreshTokenParam) {
  if (refreshTokenParam.newAuthToken) {
    let ret = {
      auth: {
        token: refreshTokenParam.newAuthToken,
        type: refreshTokenParam.newAuthTokenType ? refreshTokenParam.newAuthTokenType : 'Bearer'
      }
    };
    if (refreshTokenParam.newAuthUserState) {
      ret = {
        ...ret,
        userState: refreshTokenParam.newAuthUserState
      };
    }
    if (refreshTokenParam.newRefreshToken) {
      ret = {
        ...ret,
        refresh: refreshTokenParam.newRefreshToken
      };
    }
    return ret;
  } else if (refreshTokenParam.newRefreshToken) {
    const ret = {
      refresh: refreshTokenParam.newRefreshToken
    };
    return ret;
  } else {
    return {
      auth: null,
      refresh: null
    };
  }
}
function doSignOut() {
  return {
    auth: null
  };
}
